### DISCUSSION TITLE
Coder v2: Mange multiple workspaces, IDEs, and users

### DISCUSSION THREAD

**Discussion by bpmct** [coder/code-server#5140] [Announcements]
We're building on an open-source project to manage many remote workspaces. Edit: It is now live on https://github.com/coder/coder

Features:

- Develop over SSH, VS Code Remote, Jetbrains, Jupyter, and code-server
- Create templates for different types of workspaces, share with multiple users
- Onboard users. Users can create multiple workspaces
- Provision environments on VMs, Kubernetes, Docker, AWS, Windows, (anything on [Terraform](https://registry.terraform.io/browse/providers))
- Auto shutdown (or delete) workspaces when they are not in use
- Self-hosted (just like code-server)

The project is still early and feedback is welcome. Feel free to say hi [on Discord](https://discord.gg/coder) and share your use cases.

**Comment by bpmct** [comment]
Feel free to comment here if you have any questions too. Here's another example that provisions multiple containers as a "workspace" for microservice development:

![FRcWa0_WQAA_Xx3](https://user-images.githubusercontent.com/22407953/165800600-84d8f0b9-fb84-4593-b38e-41a7e94cf54f.jpeg)

**Reply by Nader822** [reply]
- [x] **[]()**

**Comment by Wuzado** [comment]
Congratulations on the new project announcement!

Small nitpick: `t3.medium` is listed as having `(4 cores, 4 GB of RAM)` in the "Create a personalized workspace" section, while it should be listed as having `(2 cores, 4 GB of RAM)`.

**Reply by bpmct** [reply]
Yikes, good find! I'll update that right away.

**Reply by bpmct** [reply]
Something interesting is those values are editable by the admin, not a Coder-specific feature. So the admin can decide what resources/infrastructure the developers can spin up. 

We're still looking at ways to make that more close

**Reply by Wuzado** [reply]
Does that imply easy extensibility regarding various hosting providers (niche cloud providers, private cloud)? That would be really interesting to see.

**Reply by bpmct** [reply]
Correct - workspaces are defined via TF so anything in the [Terraform registry](https://registry.terraform.io/) can be automatically be provisioned by Coder.

With that being said, non-supported cloud providers can also be used as long as you start the Coder agent on the resource with a valid token :)

**Reply by Nader822** [reply]
> Does that imply easy extensibility regarding various hosting providers (niche cloud providers, private cloud)? That would be really interesting to see.

**Comment by iamzaidsoomro** [comment]
Congratulations on the new announcement! 🖤

**Comment by lg-nineoneninethreenine** [comment]
How can you install Coder Server on a Raspberry Pi? If that's possible, I can use my own dev machines than pay for GCP.

**Reply by bpmct** [reply]
Yep! You could do it via Kubernetes or manually running the agent on each pi :)

**Reply by lg-nineoneninethreenine** [reply]
Well, I have one 8 GB Pi4B, so if I gain early access, I can run the Coder agent via that and access it on Windows 7 via Coder CLI?

**Reply by bpmct** [reply]
Correct. I have a few raspberry pis so I'll see if I can get a demo up in the docs soon

**Reply by lg-nineoneninethreenine** [reply]
10-4. I gave my name for joining!

**Comment by unknowndevQwQ** [comment]
Can I install it in npm?

**Reply by lg-nineoneninethreenine** [reply]
No, you can't. Unlike code-server, which is built with TypeScript, Coder v2 is made with the Go programming language. So, the only way to instal this is through a prebuilt release binary or building from source code.

**Reply by unknowndevQwQ** [reply]
> No, you can't. Unlike code-server, which is built with TypeScript, Coder v2 is made with the Go programming language. So, the only way to instal this is through a prebuilt release binary or building from source code.

Has support for freebsd been considered?

**Reply by jsjoeio** [reply]
I'm not sure 🤔 You can try it and see!

**Reply by lg-nineoneninethreenine** [reply]
You can build coder on freebsd, right? Just get Nix, install deps via that and run `make build`, right?

**Reply by unknowndevQwQ** [reply]
> You can build coder on freebsd, right? Just get Nix, install deps via that and run `make build`, right?

get nix package manager?

**Reply by lg-nineoneninethreenine** [reply]
Yep, that's what was given on the documentation. There would be no stupid comments like "It doesn't work on my dumb Linux installation."(could be me haha) Rather, it works universally, so A+ to use that to regulate packages, apps, etc.

**Reply by unknowndevQwQ** [reply]
> Yep, that's what was given on the documentation. There would be no stupid comments like "It doesn't work on my dumb Linux installation."(could be me haha) Rather, it works universally, so A+ to use that to regulate packages, apps, etc.

`./install: sorry, there is no binary distribution of Nix for your platform`

**Reply by lg-nineoneninethreenine** [reply]
W. T. H.

Maybe you have to ask @bpmct about this. It could be possible to build on FreeBSD if the packages are good and system requirements are met, and Nix is a universal package manager, seems like it isn't made for FreeBSD yet. I think there could be more workarounds on this.

**Reply by unknowndevQwQ** [reply]
> Maybe you have to ask @bpmct about this. It could be possible to build on FreeBSD if the packages are good and system requirements are met, and Nix is a universal package manager, seems like it isn't made for FreeBSD yet. I think there could be more workarounds on this.

Who's requirements are being met?

**Reply by lg-nineoneninethreenine** [reply]
The system requirements. If it meets a set requirement of RAM, storage and CPU cores, it can run on your system.

**Reply by unknowndevQwQ** [reply]
> The system requirements. If it meets a set requirement of RAM, storage and CPU cores, it can run on your system.

https://github.com/coder/coder/issues/4516
Unfortunately, freebsd is not currently supported

**Comment by Nader822** [comment]
- [ ]

**Comment by bilogic** [comment]
Possible to run 1 instance of `code-server` and edit projects like in GitHub, i.e. just pressing the `.`?

1. Imagine my development server is hosted on https://project1.com
2. I can start working on it by going into https://project1.com/vscode
3. Eventually it would be wonderful if we could separate developers using domains, e.g. https://developer1.project1.com, https://developer2.project1.com, https://developer3.project2.com all under 1 instance of apache and code-server

I'm not sure what this feature is called, wasn't able to find anything useful with "edit like github"

**Reply by jsjoeio** [reply]
That's a neat idea! We did explore something like that previously but then shut it down. In terms of implementing it yourself, here's what comes to mind would be to:

1. deploy code-server
2. have "middleware" that stripes path or subdomain
3. clones repo on your VM 
4. redirects and opens code-server to folder https://project1.com/?folder=/home/coder/repo

**Reply by bilogic** [reply]
Thanks, but I'm curious if there is some way to restrict access between developers' code?

**Reply by jsjoeio** [reply]
Oh! Hmm...yeah I guess you'd have to put something in front like [Pomerium](https://www.pomerium.com/docs/guides/code-server).

The _easier_ solution IMO would be to use [coder/coder](https://github.com/coder/coder) which is built for teams. Each person could get their own workspace which can have code-server.

**Reply by bilogic** [reply]
Yes, but I thought preventing access to another developers files should not be that challenging. Haha.

Not to mention the extra resources needed to run multiple code servers.

**Reply by jsjoeio** [reply]
> Yes, but I thought preventing access to another developers files should not be that challenging. Haha.

I haven't actually tried writing code to solve this so it could be easy. Hope you figure something out!

Let us know if you need more help!

**Reply by bilogic** [reply]
I gave it some thought.

1. Since `code-server` is already a HTTP server by itself, finer grain control of the folders probably needs to be implemented in `code-server` and not Pomerium/SSOs/etc
2. We could use the identity from SSO and specify a list of folders allowed to access in a JSON which is read by `code-server`, i.e.
    ```JSON
    {
        "dev1@example.com": [
            "/var/www/dev1/", // and all children
            "/home/dev1/", // and all children
        ]
    }
    ```
3. I haven't worked out how the identity can be passed from SSO to `code-server`

**Reply by jsjoeio** [reply]
That definitely seems like it could work! @code-asher might be able to share additional thoughts if he has any.

**Reply by bilogic** [reply]
I just realized you guys are using `config.yaml` and have a password-only login screen.
I'm trying to find an existing flow that we could utilize here. It seems so tempting to just add a username to the login page. :)

**Reply by jsjoeio** [reply]
LOL I mean you could fork code-server and add that 😉

**Reply by bilogic** [reply]
haha would you guys accept that PR?

**Reply by jsjoeio** [reply]
Hmm...my hunch says probably not but you're welcome to open a new issue and start a discussion

**Reply by code-asher** [reply]
I think implementing a username would be tricky because code-server gives unfettered access to the system as the user that spawned code-server.  There would need to be a good amount of plumbing involved to lock things down per user.

Making code-server support multiple users could be cool but we would need to weigh that against the simplicity of just running multiple instances which leans on native permissions.

**Reply by code-asher** [reply]
If someone does tackle this I think the things to look out for are any process spawns (terminal, extension host, etc) and HTTP file access endpoints.

**Reply by code-asher** [reply]
If Code was split out into a separate process like we used to have it this would be quite a bit easier because you could just spawn Code as whatever user you are logged in as.  Then all you have to do is check code-server's HTTP endpoints to make sure no unauthorized access is possible across different users.

Currently Code runs in the same process so if we kept that architecture a bunch of patching to Code would need to be done which would be unfortunate.

**Reply by bilogic** [reply]
@code-asher thanks for the explanation, now I understand the design concerns. And I gave it some thought.

1. I agree adding usernames can be a rabbit hole, because the next request is probably permission management, especially when my request started out asking how I can assign different folders to different developers.
2. I'm actually still undecided whether to have single instance, or multiple. Each developer would need to run PHP XDebug which might be impossible under single instance.
3. But I'm curious, why would developers require different levels of OS access? The ability to deploy? Perhaps we can run code-server with the most basic developer permissions, those who need higher access would have to `su` from the terminal

Thank you.

**Reply by bilogic** [reply]
Alternatively, is there a way to start code-server instances on demand when loading via URL? Ultimately my concern with multiple instances boils down to optimizing server resources.

**Reply by code-asher** [reply]
Good point on the permissions!  That could balloon fast.  We could try to mitigate that by using native permissions and users but that might only work well on Linux.

On point 3 do you mean creating multiple users on the system vs a single user?  It could be nice to have a user on the system for each developer to prevent them from accessing each other's files, settings, processes, etc.  That would also let you control who can use `sudo` using the sudoers file.  When you start code-server with each user it would naturally inherit their permissions.

Is the idle resource usage of code-server pretty high?  I think we only start Code when a request is made so I would expect it to be fairly low but we might have a bug there.  When disconnected Code should stop extensions and eventually stop terminals so resource usage should be fairly low there as well but again there could be bugs or maybe it is not well optimized.  We might also have leaks somewhere.

If you want to start code-server itself on demand I think you will have to write something to do that.  I have seen some people around here say they wrote their own but no one has shared code yet as far as I remember.

**Reply by bilogic** [reply]
Is there a way to run `code-server` as dev1, but prevent dev1 from mucking around with their `config.yaml`?
There are certain functions like  `--disable-file-downloads` that I wish to enforce

**Reply by code-asher** [reply]
Cross-posting in case someone finds this and wants to know the answer: https://github.com/coder/code-server/discussions/5856

**Comment by KaKi87** [comment]
> Feel free to say hi on Discord

What's funny is the README says *join us on Slack* :joy:

**Reply by code-asher** [reply]
Good point, we should probably remove that.  I think the Slack channel is more or less dead.  I never check it, at least.  :stuck_out_tongue:   I am not on Discord much either though.
